File: src/mtc_pipeline/src/end_effector_stages.cpp
#include "mtc_pipeline/end_effector_stages.hpp"
#include <moveit/task_constructor/stages/move_to.h>

namespace {
// Derives MoveIt group name from gripper type
std::string get_gripper_group(const std::string& type) {
    if (type == "pipettor") return "";  // No movable joints
    return type + "_gripper";  // hande -> hande_gripper, epick -> epick_gripper
}
}  // namespace

EndEffectorStages::EndEffectorStages(const rclcpp::Node::SharedPtr& node)
    : BaseStages(node) {}

bool EndEffectorStages::run(const mtc_pipeline::action::EndEffectorAction::Goal& goal)
{
    std::string group = get_gripper_group(goal.end_effector_type);
    if (group.empty()) {
        RCLCPP_ERROR(node()->get_logger(), "Unknown end effector: %s", goal.end_effector_type.c_str());
        return false;
    }

    RCLCPP_INFO(node()->get_logger(), "End effector: %s -> %s",
        goal.end_effector_type.c_str(), goal.end_effector_action.c_str());

    auto task = create_task_template(goal.end_effector_action, group);
    auto stage = std::make_unique<mtc::stages::MoveTo>(goal.end_effector_action, make_joint_interpolation_planner());
    stage->setGroup(group);
    stage->setGoal(goal.end_effector_action);  // Action name is the SRDF state
    task.add(std::move(stage));

    return load_plan_execute(task);
}

File: src/mtc_py/mtc_py_lib/stages/end_effector_stages.py
"""EndEffector stages - Python equivalent of end_effector_stages.hpp/cpp.

Handles gripper open/close operations using SRDF-defined group states.
"""

from moveit.task_constructor import stages
from mtc_py_lib.stages.base_stages import BaseStages


class EndEffectorStages(BaseStages):
    """Handles gripper open/close operations."""

    def run(self, goal) -> bool:
        """Execute EndEffector action.

        Args:
            goal: EndEffectorAction.Goal with fields:
                - gripper_group: MoveIt group name (from config)
                - end_effector_action: SRDF state name (e.g., "hande_open")

        Returns:
            True if successful, False otherwise
        """
        if not goal.gripper_group:
            self.logger.info("No gripper group - treating as no-op success")
            return True  # No-op success for "none" or "pipettor" gripper

        # Validate we have an action to perform
        if not goal.end_effector_action:
            self.logger.error("No end_effector_action specified")
            return False

        task = self.create_task_template("EndEffector Task")
        planner = self.make_joint_interpolation_planner()

        # Create MoveTo stage for gripper
        stage = stages.MoveTo(f"gripper_{goal.end_effector_action}", planner)
        stage.group = goal.gripper_group
        stage.setGoal(goal.end_effector_action)

        task.add(stage)

        self.logger.info(
            f"Planning gripper action: {goal.end_effector_action} "
            f"(group: {goal.gripper_group})"
        )
        return self.load_plan_execute(task)
